Sources/Datadog/Core/Upload/DataUploadStatus.swift

```diff
@@ -0,0 +1,164 @@
+            internalMonitoringError: createInternalMonitoringErrorIfNeeded(for: httpResponse.statusCode, requestID: ddRequestID)
+        )
+    }
+
+    init(networkError: Error) {
+        self.init(
+            needsRetry: true, // retry this upload as it failed due to network transport isse
+            userDebugDescription: "[error: \(DDError(error: networkError).message)]", // e.g. "[error: A data connection is not currently allowed]"
+            userErrorMessage: nil, // nothing actionable for the user
+            internalMonitoringError: createInternalMonitoringErrorIfNeeded(for: networkError)
+        )
+    }
+}
+
+// MARK: - Internal Monitoring
+
+#if DD_SDK_ENABLE_INTERNAL_MONITORING
+/// Looks at the `statusCode` and produces error for Internal Monitoring feature if anything is going wrong.
+private func createInternalMonitoringErrorIfNeeded(
+    for statusCode: Int, requestID: String?
+) -> (message: String, error: Error?, attributes: [String: String]?)? {
+    guard let responseStatusCode = HTTPResponseStatusCode(rawValue: statusCode) else {
+        // If status code is unexpected, do not produce an error for Internal Monitoring - otherwise monitoring may
+        // become too verbose for old installations if we introduce a new status code in the API.
+        return nil
+    }
+
+    switch responseStatusCode {
+    case .accepted, .unauthorized, .forbidden:
+        // These codes mean either success or the user configuration mistake - do not produce error.
+        return nil
+    case .tooManyRequests, .internalServerError, .serviceUnavailable:
+        // These codes mean Datadog service issue - do not produce SDK error as this is already monitored by other means.
+        return nil
+    case .badRequest, .payloadTooLarge, .requestTimeout:
+        // These codes mean that something wrong is happening either in the SDK or on the server - produce an error.
+        return (
+            message: "Data upload finished with status code: \(statusCode)",
+            error: nil,
+            attributes: ["dd_request_id": requestID ?? "(???)"]
+        )
+    case .unexpected:
+        return nil
+    }
+}
+
+/// A list of known NSURLError codes which should not produce error in Internal Monitoring.
+/// Receiving these codes doesn't mean SDK issue, but the network transportation scenario where the connection interrupted due to external factors.
+/// These list should evolve and we may want to add more codes in there.
+///
+/// Ref.: https://developer.apple.com/documentation/foundation/1508628-url_loading_system_error_codes
+private let ignoredNSURLErrorCodes = Set([
+    NSURLErrorNetworkConnectionLost, // -1005
+    NSURLErrorTimedOut, // -1001
+    NSURLErrorCannotParseResponse, // - 1017
+    NSURLErrorNotConnectedToInternet, // -1009
+    NSURLErrorCannotFindHost, // -1003
+    NSURLErrorSecureConnectionFailed, // -1200
+    NSURLErrorDataNotAllowed, // -1020
+    NSURLErrorCannotConnectToHost, // -1004
+])
```

Comment:
I made this list by looking at Datadog iOS app telemetry. None of these codes looks like SDK fault:
<img width="1396" alt="Screenshot 2021-08-12 at 14 35 40" src="https://user-images.githubusercontent.com/2358722/129197740-eac3a3c5-7351-46a3-8967-d94ae099a4a5.png">